src/main/java/org/quiltmc/loader/impl/discovery/ModResolver.java

```diff
@@ -551,6 +731,180 @@ private static boolean isModIdValid(String modId, List<String> errorList) {
+
+						for (ModLoadOption source : sources) {
+							errors.append("\n\t\t- ").append(getCandidateFriendlyVersion(source)).append(".");
+						}
+					}
+				}
+			} else if (cause instanceof ModBreakage) {
+				ModBreakage breakage = (ModBreakage) cause;
+				errors.append(breakage.invalidOptions.isEmpty() ? "-" : "x");
+				errors.append(" Mod ").append(getLoadOptionDescription(breakage.source))
+						.append(" conflicts with ").append(getDependencyVersionRequirements(breakage.publicDep))
+						.append(" of ");
+
+				ModIdDefinition def = breakage.with;
+				ModLoadOption[] sources = def.sources();
+
+				if (sources.length == 0) {
+					errors.append("unknown mod '").append(def.getModId()).append("'\n")
+							.append("\t- You must remove ").append(getDependencyVersionRequirements(breakage.publicDep))
+							.append(" of '").append(def.getModId()).append("'.");
+				} else {
+					errors.append(def.getFriendlyName());
+
+					if (breakage.invalidOptions.isEmpty()) {
+						errors.append("\n\t- You must remove ").append(getDependencyVersionRequirements(breakage.publicDep))
+								.append(" of ").append(def.getFriendlyName()).append('.');
+					}
+
+					if (sources.length == 1) {
+						errors.append("\n\t- Your current version of ").append(getCandidateName(sources[0].candidate))
+							.append(" is ").append(getCandidateFriendlyVersion(sources[0].candidate)).append(".");
+					} else {
+						errors.append("\n\t- You have the following versions available:");
+
+						for (ModLoadOption source : sources) {
+							errors.append("\n\t\t- ").append(getCandidateFriendlyVersion(source)).append(".");
+						}
+					}
+				}
+			} else {
+				errors.append("x Unknown error type?")
+						.append("\n\t+ cause.getClass() =>")
+						.append("\n\t\t").append(cause.getClass().getName())
+						.append("\n\t+ cause.toString() =>")
+						.append("\n\t\t").append(cause.toString());
+			}
+		}
+
+		// TODO: See if I can get results similar to appendJiJInfo (which requires a complete "mod ID -> candidate" map)
+		HashSet<String> listedSources = new HashSet<>();
+		for (ModLoadOption involvedMod : roots.keySet()) {
+			appendLoadSourceInfo(errors, listedSources, involvedMod);
+		}
+
+		for (ModLink involvedLink : causes) {
+			if (involvedLink instanceof ModDep) {
+				appendLoadSourceInfo(errors, listedSources, ((ModDep) involvedLink).on);
+			} else if (involvedLink instanceof ModBreakage) {
+				appendLoadSourceInfo(errors, listedSources, ((ModBreakage) involvedLink).with);
+			}
+		}
+
+		return new ModResolutionException(errors.toString());
+	}
+
+	private static void appendLoadSourceInfo(StringBuilder errors, HashSet<String> listedSources, ModIdDefinition def) {
+		if (!listedSources.add(def.getModId())) {
+			return;
+		}
+
+		ModLoadOption[] sources = def.sources();
+
+		if (sources.length == 0) {
+			return;
+		}
+
+		if (sources.length == 1) {
+			errors.append("\n- $jar+fabric$ ").append(getLoadOptionDescription(sources[0]))
```

Comment:
My original intention was that we'd use the "fabric" icon to indicate a "fabric.mod.json" based mod, and the quilt icon for "quilt.mod.json" based mods. As QMJ isn't finished quite yet all mods are fabric based, although I do need to move this behaviour to ModLoadOption / ModCandidate rather than hardcode them here.